go/service/chat_boxer.go

```diff
@@ -0,0 +1,160 @@
+		if key.KeyGeneration == msg.KeyGeneration {
+			matchKey = &key
+			break
+		}
+	}
+
+	if matchKey == nil {
+		return keybase1.Message{}, fmt.Errorf("no key found for generation %d", msg.KeyGeneration)
+	}
+
+	return b.unboxMessageWithKey(msg, matchKey)
+}
+
+// unboxMessageWithKey unboxes a chat1.MessageBoxed into a keybase1.Message given
+// a keybase1.CryptKey.
+func (b *chatBoxer) unboxMessageWithKey(msg chat1.MessageBoxed, key *keybase1.CryptKey) (keybase1.Message, error) {
+	if msg.ServerHeader == nil {
+		return keybase1.Message{}, errors.New("nil ServerHeader in MessageBoxed")
+	}
+
+	if err := b.verifyMessageBoxed(msg); err != nil {
+		return keybase1.Message{}, err
+	}
+
+	unboxed := keybase1.Message{
+		ServerHeader: *msg.ServerHeader,
+	}
+
+	if len(msg.BodyCiphertext.N) != libkb.NaclDHNonceSize {
+		return keybase1.Message{}, libkb.DecryptBadNonceError{}
+	}
+	var nonce [libkb.NaclDHNonceSize]byte
+	copy(nonce[:], msg.BodyCiphertext.N)
+
+	plaintext, ok := secretbox.Open(nil, msg.BodyCiphertext.E, &nonce, ((*[32]byte)(&key.Key)))
+	if !ok {
+		return keybase1.Message{}, libkb.DecryptOpenError{}
+	}
+
+	if err := json.Unmarshal(plaintext, &unboxed.MessagePlaintext); err != nil {
+		return keybase1.Message{}, err
+	}
+
+	return unboxed, nil
+
+}
+
+// boxMessage encrypts a keybase1.MessagePlaintext into a chat1.MessageBoxed.  It
+// finds the most recent key for the TLF.
+func (b *chatBoxer) boxMessage(ctx context.Context, msg keybase1.MessagePlaintext) (chat1.MessageBoxed, error) {
+	tlfName := msg.ClientHeader.TlfName
+	keys, err := b.tlf.CryptKeys(ctx, tlfName)
+	if err != nil {
+		return chat1.MessageBoxed{}, err
+	}
+
+	var recentKey *keybase1.CryptKey
+	for _, key := range keys.CryptKeys {
+		if recentKey == nil || key.KeyGeneration > recentKey.KeyGeneration {
+			recentKey = &key
+		}
+	}
+
+	if recentKey == nil {
+		return chat1.MessageBoxed{}, fmt.Errorf("no key found for tlf %q", tlfName)
+	}
+
+	return b.boxMessageWithKey(msg, recentKey)
+}
+
+// boxMessageWithKey encrypts a keybase1.MessagePlaintext into a chat1.MessageBoxed
+// given a keybase1.CryptKey.
+func (b *chatBoxer) boxMessageWithKey(msg keybase1.MessagePlaintext, key *keybase1.CryptKey) (chat1.MessageBoxed, error) {
+	s, err := json.Marshal(msg)
+	if err != nil {
+		return chat1.MessageBoxed{}, err
+	}
+
+	var nonce [libkb.NaclDHNonceSize]byte
+	if _, err := rand.Read(nonce[:]); err != nil {
+		return chat1.MessageBoxed{}, err
+	}
+
+	sealed := secretbox.Seal(nil, []byte(s), &nonce, ((*[32]byte)(&key.Key)))
+
+	boxed := chat1.MessageBoxed{
+		ClientHeader: msg.ClientHeader,
+		BodyCiphertext: chat1.EncryptedData{
+			V: 1,
+			E: sealed,
+			N: nonce[:],
+		},
+		KeyGeneration: key.KeyGeneration,
+	}
+
+	return boxed, nil
+}
+
+// verifyMessageBoxed verifies the header and body signatures in a boxed
+// message.
+func (b *chatBoxer) verifyMessageBoxed(msg chat1.MessageBoxed) error {
+	header, err := json.Marshal(msg.ClientHeader)
```

Comment:
checks I couldn't find:

* Is it the same public key in both cases (it should be)
* Is the public key valid for the user (or was it valid at the time the signature was generated (which we can't really trust since it can be backdated))
* does the header in `msg.ClientHeader` match the header in the `msg.BodyCiphertext.E`
* does the header have the right fields relative to this conversation, such as the prev pointers you mentioned above, but also the right TLF ID, etc.  Also, is the time stamp reasonable?